drivers/usb/bc12/bc12_pi3usb9201.c

```diff
@@ -0,0 +1,432 @@
+#include "bc12_pi3usb9201.h"
+
+LOG_MODULE_REGISTER(PI3USB9201, CONFIG_USB_BC12_LOG_LEVEL);
+
+/* Constant configuration data */
+struct pi3usb9201_config {
+	struct i2c_dt_spec i2c;
+
+	struct gpio_dt_spec intb_gpio;
+
+	enum bc12_type charging_mode;
+};
+
+/* Run-time configuration data */
+struct pi3usb9201_data {
+	const struct device *dev;
+	struct k_work work;
+	enum bc12_type partner_type;
+	struct gpio_callback gpio_cb;
+
+	bc12_callback_t result_cb;
+	void *result_cb_data;
+};
+
+enum pi3usb9201_client_sts {
+	CHG_OTHER = 0,
+	CHG_2_4A = 1,
+	CHG_2_0A = 2,
+	CHG_1_0A = 3,
+	CHG_RESERVED = 4,
+	CHG_CDP = 5,
+	CHG_SDP = 6,
+	CHG_DCP = 7,
+};
+
+struct bc12_status {
+	enum bc12_type partner_type;
+	int current_limit;
+};
+
+/*
+ * The USB Type-C specification limits the maximum amount of current from BC 1.2
+ * suppliers to 1.5A.  Technically, proprietary methods are not allowed, but we
+ * will continue to allow those.
+ */
+static const struct bc12_status bc12_chg_limits[] = {
+	/* For unknown chargers return Isusp. */
+	[CHG_OTHER] = {BC12_TYPE_PROPRIETARY, BC12_CURR_UA(BC12_CHARGER_MIN_CURR_UA)},
+	[CHG_2_4A] = {BC12_TYPE_PROPRIETARY, BC12_CURR_UA(2400000)},
+	[CHG_2_0A] = {BC12_TYPE_PROPRIETARY, BC12_CURR_UA(2000000)},
+	[CHG_1_0A] = {BC12_TYPE_PROPRIETARY, BC12_CURR_UA(1000000)},
+	[CHG_RESERVED] = {BC12_TYPE_NONE, 0},
+	[CHG_CDP] = {BC12_TYPE_CDP, BC12_CURR_UA(1500000)},
+	/* BC1.2 driver contract specifies to return Isusp for SDP ports. */
+	[CHG_SDP] = {BC12_TYPE_SDP, BC12_CURR_UA(BC12_CHARGER_MIN_CURR_UA)},
+	[CHG_DCP] = {BC12_TYPE_DCP, BC12_CURR_UA(1500000)},
+};
+
+static int pi3usb9201_interrupt_enable(const struct device *dev, const bool enable)
+{
+	const struct pi3usb9201_config *cfg = dev->config;
+
+	/* Clear the interrupt mask bit to enable the interrupt */
+	return i2c_reg_update_byte_dt(&cfg->i2c, PI3USB9201_REG_CTRL_1,
+				      PI3USB9201_REG_CTRL_1_INT_MASK,
+				      enable ? 0 : PI3USB9201_REG_CTRL_1_INT_MASK);
+}
+
+static int pi3usb9201_bc12_detect_ctrl(const struct device *dev, const bool enable)
+{
+	const struct pi3usb9201_config *cfg = dev->config;
+
+	return i2c_reg_update_byte_dt(&cfg->i2c, PI3USB9201_REG_CTRL_2,
+				      PI3USB9201_REG_CTRL_2_START_DET,
+				      enable ? PI3USB9201_REG_CTRL_2_START_DET : 0);
+}
+
+static int pi3usb9201_bc12_usb_switch(const struct device *dev, bool enable)
+{
+	const struct pi3usb9201_config *cfg = dev->config;
+
+	/* USB data switch enabled when PI3USB9201_REG_CTRL_2_AUTO_SW is clear */
+	return i2c_reg_update_byte_dt(&cfg->i2c, PI3USB9201_REG_CTRL_2,
+				      PI3USB9201_REG_CTRL_2_START_DET,
+				      enable ? 0 : PI3USB9201_REG_CTRL_2_AUTO_SW);
+}
+
+static int pi3usb9201_set_mode(const struct device *dev, enum pi3usb9201_mode mode)
+{
+	const struct pi3usb9201_config *cfg = dev->config;
+
+	return i2c_reg_update_byte_dt(&cfg->i2c, PI3USB9201_REG_CTRL_1,
+				      PI3USB9201_REG_CTRL_1_MODE_MASK
+					      << PI3USB9201_REG_CTRL_1_MODE_SHIFT,
```

Comment:
I'm not sure if you particularly need to have the mask and shift as separate values, but I found `FIELD_PREP()` from `<zephyr/sys/util.h>` very handy so I can just create a mask via `GENMASK()` and set the values without worrying about the extra shifting and masking.